### PR TITLE
Change immutable field errors to type Forbidden

### PR DESCRIPTION
**Pull request by mitalipaygude**
*Issue #, if available:*

*Description of changes:*
Change immutable field errors to type Forbidden instead of Invalid so as to follow CAPI.
Reference: https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/machinehealthcheck_webhook.go#L135

*Testing (if applicable):*
- Unit tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go [hunk 1]
```diff
@@ -87,7 +87,7 @@ func validateImmutableFieldsCloudStackCluster(new, old *CloudStackDatacenterConf
 	if old.Spec.Domain != new.Spec.Domain {
 		allErrs = append(
 			allErrs,
-			field.Invalid(field.NewPath("spec", "domain"), new.Spec.Domain, "field is immutable"),
+			field.Forbidden(field.NewPath("spec", "domain", new.Spec.Domain), "field is immutable"),
```
Comment (maxdrib) [inline pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go:90]: What's the motivation for this change?

Comment (mitalipaygude) [reply pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go:90]: CAPI uses forbidden for immutable fields, instead of Invalid. Invalid fields can generally mean its a string instead of int, but we want to forbid changing the value at all for immutable fields.

Comment (maxdrib) [reply pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go:90]: Gotcha. In that case we should be good from Cloudstack, just need to resolve the conflicts

File: pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go [hunk 2]
```diff
@@ -104,21 +104,21 @@ func validateImmutableFieldsCloudStackCluster(new, old *CloudStackDatacenterConf
 	if zonesMutated {
 		allErrs = append(
 			allErrs,
-			field.Invalid(field.NewPath("spec", "zone"), new.Spec.Zones, "field is immutable"),
+			field.Forbidden(field.NewPath("spec", "zone"), "field is immutable"),
```
Comment (vivek-koppuru) [inline pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go:107]: Should this be the following? But check the spec fields in the api package
```suggestion
			field.Forbidden(field.NewPath("spec", "zones"), "field is immutable"),
```

File: pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go [hunk 3]
```diff
@@ -104,21 +104,21 @@ func validateImmutableFieldsCloudStackCluster(new, old *CloudStackDatacenterConf
...
 		)
 	}
 
 	if old.Spec.Account != new.Spec.Account {
 		allErrs = append(
 			allErrs,
-			field.Invalid(field.NewPath("spec", "account"), new.Spec.Account, "field is immutable"),
+			field.Forbidden(field.NewPath("spec", "account", new.Spec.Account), "field is immutable"),
```
Comment (vivek-koppuru) [inline pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go:114]: ?
```suggestion
			field.Forbidden(field.NewPath("spec", "account"), "field is immutable"),
```

File: pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go [hunk 4]
```diff
@@ -104,21 +104,21 @@ func validateImmutableFieldsCloudStackCluster(new, old *CloudStackDatacenterConf
...
 		)
 	}
 
 	if old.Spec.ManagementApiEndpoint != new.Spec.ManagementApiEndpoint {
 		allErrs = append(
 			allErrs,
-			field.Invalid(field.NewPath("spec", "managementApiEndpoint"), new.Spec.ManagementApiEndpoint, "field is immutable"),
+			field.Forbidden(field.NewPath("spec", "managementApiEndpoint", new.Spec.ManagementApiEndpoint), "field is immutable"),
```
Comment (vivek-koppuru) [inline pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go:121]: ?
```suggestion
			field.Forbidden(field.NewPath("spec", "managementApiEndpoint"), "field is immutable"),
```

File: pkg/api/v1alpha1/awsiamconfig_webhook.go [hunk 5]
```diff
@@ -78,7 +78,7 @@ func validateImmutableAWSIamFields(new, old *AWSIamConfig) field.ErrorList {
 	if !new.Spec.Equal(&old.Spec) {
 		allErrs = append(
 			allErrs,
-			field.Invalid(field.NewPath("spec", "AWSIamConfig"), new, "config is immutable"),
+			field.Forbidden(field.NewPath("spec", "AWSIamConfig", new.Name), "config is immutable"),
```
Comment (vivek-koppuru) [inline pkg/api/v1alpha1/awsiamconfig_webhook.go:81]: I wonder if this is better? However we should see what capi does as an example
```suggestion
			field.Forbidden(field.NewPath("AWSIamConfig", new.Name), "spec is immutable"),
```

File: pkg/api/v1alpha1/fluxconfig_webhook.go [hunk 6]
```diff
@@ -90,7 +90,7 @@ func validateImmutableFluxFields(new, old *FluxConfig) field.ErrorList {
 	if !new.Spec.Equal(&old.Spec) {
 		allErrs = append(
 			allErrs,
-			field.Invalid(field.NewPath("spec", FluxConfigKind), new, "config is immutable"),
+			field.Forbidden(field.NewPath("spec", FluxConfigKind, new.Name), "config is immutable"),
```
Comment (vivek-koppuru) [inline pkg/api/v1alpha1/fluxconfig_webhook.go:93]: Same as above comment for `AWSIamConfig`, as well as all the other files below

File: pkg/api/v1alpha1/vspheredatacenterconfig_webhook.go [hunk 7]
```diff
@@ -93,35 +93,35 @@ func validateImmutableFieldsVSphereCluster(new, old *VSphereDatacenterConfig) fi
 	if old.Spec.Server != new.Spec.Server {
 		allErrs = append(
 			allErrs,
-			field.Invalid(field.NewPath("spec", "server"), new.Spec.Server, "field is immutable"),
+			field.Forbidden(field.NewPath("spec", "server", new.Spec.Server), "field is immutable"),
```
Comment (vivek-koppuru) [inline pkg/api/v1alpha1/vspheredatacenterconfig_webhook.go:96]: Same goes here on whether we want to remove that field if we are already specifying "server", as well as the ones below in this file.

File: pkg/api/v1alpha1/awsiamconfig_webhook.go [hunk 8]
```diff
@@ -78,7 +78,7 @@ func validateImmutableAWSIamFields(new, old *AWSIamConfig) field.ErrorList {
 	if !new.Spec.Equal(&old.Spec) {
 		allErrs = append(
 			allErrs,
-			field.Invalid(field.NewPath("spec", "AWSIamConfig"), new, "config is immutable"),
+			field.Forbidden(field.NewPath("AWSIamConfig", new.Name), "config is immutable"),
```
Comment (g-gaston) [inline pkg/api/v1alpha1/awsiamconfig_webhook.go:81]: What is the logic we follow to define the path?
It seems like we are removing spec from all of them, was that redundant?
Also, why is `new.Name` part of the path?

File: pkg/api/v1alpha1/cluster_webhook.go [hunk 9]
```diff
@@ -113,56 +113,57 @@ func validateImmutableFieldsCluster(new, old *Cluster) field.ErrorList {
...
 			allErrs,
-			field.Forbidden(field.NewPath("spec", "ProxyConfiguration"), fmt.Sprintf("field is immutable %v", new.Spec.ProxyConfiguration)))
+			field.Forbidden(specPath.Child("ProxyConfiguration"), fmt.Sprintf("field is immutable %v", new.Spec.ProxyConfiguration)))
 	}
 
 	if new.Spec.ExternalEtcdConfiguration != nil && old.Spec.ExternalEtcdConfiguration == nil {
 		allErrs = append(
 			allErrs,
-			field.Forbidden(field.NewPath("spec.externalEtcdConfiguration"), "cannot switch from local to external etcd topology"),
+			field.Forbidden(specPath.Child("spec.externalEtcdConfiguration"), "cannot switch from local to external etcd topology"),
```
Comment (g-gaston) [inline pkg/api/v1alpha1/cluster_webhook.go:151]: this already contains spec

File: pkg/api/v1alpha1/awsiamconfig_webhook.go [hunk 10]
```diff
@@ -74,11 +74,12 @@ func (r *AWSIamConfig) ValidateDelete() error {
 
 func validateImmutableAWSIamFields(new, old *AWSIamConfig) field.ErrorList {
 	var allErrs field.ErrorList
+	specPath := field.NewPath("spec")
 
 	if !new.Spec.Equal(&old.Spec) {
 		allErrs = append(
 			allErrs,
-			field.Invalid(field.NewPath("spec", "AWSIamConfig"), new, "config is immutable"),
+			field.Forbidden(specPath.Child("AWSIamConfig"), "config is immutable"),
```
Comment (vivek-koppuru) [inline pkg/api/v1alpha1/awsiamconfig_webhook.go:82]: Does `spec.AWSIamConfig` make sense though? Can we somehow just say "AWSIamConfig <name>: config is immutable"?

File: pkg/api/v1alpha1/cluster_webhook.go [hunk 11]
```diff
@@ -186,7 +187,7 @@ func validateImmutableFieldsCluster(new, old *Cluster) field.ErrorList {
 		if !oldAWSIamConfig.Equal(newAWSIamConfig) {
 			allErrs = append(
 				allErrs,
-				field.Forbidden(field.NewPath("spec", "AWS Iam Config"), fmt.Sprintf("field is immutable %v", newAWSIamConfig.Kind)))
+				field.Forbidden(specPath.Child("AWS Iam Config"), fmt.Sprintf("field is immutable %v", newAWSIamConfig.Kind)))
```
Comment (vivek-koppuru) [inline pkg/api/v1alpha1/cluster_webhook.go:190]: This is not a field right? It's the entire config, unless this is referring to the ref field, which we should refer to that instead.

Comment (mitalipaygude) [reply pkg/api/v1alpha1/cluster_webhook.go:190]: Not a field technically, its a ref. But when users edit it, we treat it as a field that is a reference.

File: pkg/api/v1alpha1/awsiamconfig_webhook.go (hunk 10 above)
Comment (mitalipaygude) [reply pkg/api/v1alpha1/awsiamconfig_webhook.go:82]: So capi adds the spec path to all these fields we call out, so to stick with capi, we add spec as well. Which means the entire AWS IAM config spec is not mutable, with all the sub fields.

File: pkg/api/v1alpha1/cluster_webhook.go (hunk 11 above)
Comment (vivek-koppuru) [reply pkg/api/v1alpha1/cluster_webhook.go:190]:
```suggestion
				field.Forbidden(specPath.Child("identityProviderRefs", "kind", "AWSIamConfig"), fmt.Sprintf("field is immutable %v", newAWSIamConfig.Kind)))
```

**Review by vivek-koppuru — APPROVED** [review]

File: pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go [hunk 12]
```diff
@@ -124,7 +125,7 @@ func validateImmutableFieldsCloudStackCluster(new, old *CloudStackDatacenterConf
 			if !newAz.Equal(&oldAz) {
 				allErrs = append(
 					allErrs,
-					field.Invalid(field.NewPath("spec", "availabilityZone", oldAz.Name), newAz, "availabilityZone is immutable"),
+					field.Forbidden(specPath.Child("availabilityZone", oldAz.Name), "availabilityZone is immutable"),
```
Comment (maxdrib) [inline pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go:128]: there's another field.Invalid reference lower in this file. Should that change as well?

Comment (mitalipaygude) [reply pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go:128]: that reference isnt for an immutable error/field. We are updating the Invalid to Forbidden only for immutable cases.
